### PR TITLE
Intl Grapheme: Adjust `grapheme_substr` PHP 8 compatibility

### PR DESCRIPTION
**Pull request by Ayesh**
`grapheme_substr` is changed in [PHP 8.0 to return an empty string on out-of-bound offsets](https://php.watch/versions/8.0/substr-out-of-bounds#grapheme_substr).

This commit updates the tests with `@requires` annotations and polyfill updates to make sure the return results are in line with native function behavior. In PHP 8, an empty string (`""`) is returned if the length parameter is out of bounds of the given string. Prior to PHP 8.0, boolean `false` is returned.

Related: #288

### CODE REVIEW

**Comment by Ayesh** [comment]
Apologies about the force-push noise, some tests were failing because older PHPUnit versions apparently ignore `@requires PHP >= 8.0` style annotations, so I had to use `@requires PHP 8.0` instead.

File: src/Intl/Grapheme/Grapheme.php [hunk 1]
```diff
@@ -123,7 +123,7 @@ public static function grapheme_substr($s, $start, $len = null)
             return false;
         }
         if ($start >= $slen) {
-            return false;
+            return \PHP_VERSION_ID >= 80000 ? '' : false;
```
Comment (nicolas-grekas) [inline src/Intl/Grapheme/Grapheme.php:126]: what about L123 & L138?

Comment (Ayesh) [reply src/Intl/Grapheme/Grapheme.php:126]: Thanks a lot for the review @nicolas-grekas, you are right, those two `false` statements need to be changed to conditionally return empty strings on PHP 8 too. 

There was not enough test coverage for those edge cases, so the new commit extends the tests with them too: 
 - `grapheme_substr($c, -42, 0)`
 - `grapheme_substr($c, 42, 5)`

**Comment by Ayesh** [comment]
I sent a PR for `iconv` before, #289, that is unfortunately incomplete. I will send a new PR to cover them too. I apologize for the time wasted and the excessive noise with those PRs, and I will be more careful to pay attention to edge cases too.

**Comment by nicolas-grekas** [comment]
No worries. Please add to this very PR, that ok.

**Comment by Ayesh** [comment]
Thank you, I cherry-picked the commit from #302 and closed it.

**Comment by nicolas-grekas** [comment]
Thank you @Ayesh.
